core/logger/zap.go

```diff
@@ -65,6 +75,42 @@ func (d *AtomicCore) Write(e zapcore.Entry, fs []zapcore.Field) error { return d
 
 func (d *AtomicCore) Sync() error { return d.load().Sync() }
 
+func (d *AtomicCore) Close() {
+	close(d.stopCleanup)
+	d.cleanupWg.Wait()
+}
+
+func (d *AtomicCore) cleanup() {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.children = slices.DeleteFunc(d.children, func(p weak.Pointer[withCore]) bool {
+		c := p.Value()
+		if c == nil {
+			return true
+		}
+		c.cleanup()
```

Comment:
I think it is not necessary to hold all of the locks at the same time, but I'm not sure if that optimization is worthwhile... 🤔 
I guess a simple version could be just:
```suggestion
		go func() {
			defer wg.Done()
			c.cleanup()
		} ()
```
with a `defer wg.Wait()` up top